### ISSUE TITLE
Shopify CDN Cache issues with CSS assets

### ISSUE THREAD

**Issue by kenji-twp** [barrel/barrel-shopify#25]
Hi guys, again thanks for the amazing work on the vite-plugin-shopify.

I had css cache issues on production, and it seems it's due the changes introduced via https://github.com/barrel/barrel-shopify/commit/45a3e2896cece57c6717a1ccd4da6ecab1c80e2d (Remove querystring from asset URLs).

https://github.com/barrel/barrel-shopify/blob/8619c4a8bb73fec3bda132e8e68d214ff3eff98d/packages/vite-plugin-shopify/src/html.ts#L162

I believe this was a fix for a Vue issue (https://github.com/barrel/barrel-shopify/issues/13), but maybe there's a way to solve both the cache and the Vue problems.

At the moment I'm using patch-package to work around this issue, but I imagine other people will have problems.

**Comment by montalvomiguelo** [comment]
@kenji-twp Thanks for using our plugin and helping us improve it 🙏 

Are you modifying dist file names in some way? _Prolly I would need to take a look at your vite config file._

By default, the dist file names contain a [[hash]](https://rollupjs.org/guide/en/#outputassetfilenames) string that is generated based on the content of the asset e.g. `theme.856128a9.css`

That [hash] will help us deliver the latest version for a given asset and using the `?v=N` param would not be necessary.


https://user-images.githubusercontent.com/5134470/203487550-43f53bbe-9223-4c6b-9535-2dc650672918.mov

**Comment by slavamak** [comment]
Hey there! I think I caught the problem that @kenji-twp mentioned. Watch the video to see it with me

https://user-images.githubusercontent.com/28267670/203829981-e46b5df3-0e19-4b2d-8825-9372a86518d8.mp4

So if we don't make changes to the entry point file, the hash is not updated. But for example we change the styles, then the hash for the styles is updated but for the entry point the hash remains the same. This is strange because the content of the file has changed. I'm not sure exactly how it works, but the shopify cache doesn't update the entry point file (maybe because the file size hasn't changed) and I get an error that the new styles file isn't found. Do you know how it works, maybe I'm missing something? 

I'm not sure what the problem was with the vue components, but if for some reason we can't use ?v= from the shopify cdn, I suppose we could avoid it if we added our timestamp to the entry point files that were changed during the build. Does it make sense?

```ts
`{{ '${fileName}' | asset_url | split: '?v=' | first | append: '?v=${timestamp}' }}`
```

**Comment by montalvomiguelo** [comment]
Thanks for that @slavamak  🙏 , I can reproduce the bug on my end.

That would work. The thing is that when we leave the ?v= param, JS chunks could download twice (First, when loading them from liquid files using the asset_url filter. Then, when other JS chunks import the same file). That was the main reason why I removed the ?v param in #21

<a href="https://montalvomiguelo.myshopify.com/?preview_theme_id=124852011056" target="_blank">
<img width="1904" alt="Screen Shot 2022-11-24 at 11 35 06 PM" src="https://user-images.githubusercontent.com/5134470/203937964-e186a51b-595e-4f0a-b099-3356475c2532.png">
</a>

I think we would need a way to make the [hash] change based on the content of the file. Not sure why it's not changing right now.

**Comment by slavamak** [comment]
thanks @montalvomiguelo, yes, it looks like this is a known issue in the Vite, because I found this [#9583](https://github.com/vitejs/vite/issues/9583)

**Comment by slavamak** [comment]
I've studied it a bit. It seems we should wait for Vite [v4](https://github.com/vitejs/vite/blob/v4.0.0-alpha.5/packages/vite/CHANGELOG.md). Because they will update the Rollup version to [v3](https://github.com/rollup/rollup/pull/4543/), in which the hashing algorithm has been redesigned and more. I also played with this hook [augmentChunkHash](https://rollupjs.org/guide/en/#augmentchunkhash), and it seems to solve our problem. in fact we will update the hash of the chunk if the imported modules have been updated. But I think it can be solved more elegantly :sweat_smile: let's see

```ts
function updateChunkHash(): Plugin {
  return {
    name: 'update-chunk-hash',

    augmentChunkHash({ viteMetadata }) {
      const importedModules = Array.from([
        ...viteMetadata.importedCss,
        ...viteMetadata.importedAssets,
      ])

      if (importedModules.length) {
        return importedModules.toString()
      }
    },
  }
}
```

**Comment by montalvomiguelo** [comment]
Thanks a lot for that @slavamak 🙏🙏

I think it's a good workaround while we wait for Vite 4. Please let me know what you think about #27

**Comment by slavamak** [comment]
This works fine for me! The only thing is that with every build the hash will change even if there were no changes in the files, but I think that's ok as a temporary solution, Thank you!

**Comment by slavamak** [comment]
This seems to have been fixed in [vite 4.0.4](https://github.com/vitejs/vite/blob/v4.0.4/packages/vite/CHANGELOG.md#404-2023-01-03) :)
